gui/layers/_vectors_layer/model.py

```diff
@@ -0,0 +1,615 @@
+
+from typing import Union
+
+import numpy as np
+
+from .._base_layer import Layer
+from .._register import add_to_viewer
+from ..._vispy.scene.visuals import Line as LinesNode
+from vispy.color import get_color_names
+
+from vispy.util.event import Event
+
+from .view import QtVectorsLayer
+from .view import QtVectorsControls
+
+
+@add_to_viewer
+class Vectors(Layer):
+    """
+    Vectors layer renders lines onto the image.
+    There are currently two data modes supported:
+        1) image-like vectors where every pixel defines its own vector
+        2) coordinate-like vectors where x-y position is not fixed to a grid
+    Supports ONLY 2d vector field
+
+    Properties
+    ----------
+    vectors : np.ndarray
+        array of shape (N,4) or (N, M , 2)
+
+    data : np.ndarray
+        ABC implementation.
+        returns vectors property setter/getter
+
+    _original_data : np.ndarray
+        Used by averaging and length adjustments.
+        Is set during layer construction.
+        Is NOT updated if data is adjusted by assigning the 'vectors' property
+
+    averaging : str
+        kernel over which to average from one of _kernel_dict
+        averaging.setter adjusts the underlying data
+        subscribe an observer by registering it with "averaging_bind_to"
+
+    width : int
+        width of the line in pixels
+
+    length : int or float
+        length of the line
+        length.setter adjusts the underlying data
+        subscribe an observer by registering it with "length_bind_to"
+
+    color : str
+        one of "get_color_names" from vispy.color
+
+    conector : str or np.ndarray
+        way to render line between coordinates
+        two built in types: "segment" or "strip"
+        third type is np.ndarray
+
+    mode : str
+        control panel mode
+
+    Attributes
+    ----------
+    Private attributes
+    -------
+    _kernel_dict
+    _kernel
+    _data_types
+    _data_type
+    _width
+    _color
+    _colors
+    _connector_types
+    _connector
+    _avg_dims
+    _averaging
+    _length
+    _avg_observers
+    _len_observers
+    _need_display_update
+    _need_visual_update
+    _raw_dat
+    _original_data
+    _current_shape
+    _vectors
+    _mode
+    _mode_history
+
+    Public attributes
+    -------
+    name
+
+
+    See vispy's line visual docs for more details:
+    http://api.vispy.org/en/latest/visuals.html#vispy.visuals.LineVisual
+    http://vispy.org/visuals.html
+    """
+
+    def __init__(
+        self, vectors,
+            width=1,
+            color='red',
+            connector='segments',
+            averaging='1x1',
+            length=1):
+
+        visual = LinesNode()
+        super().__init__(visual)
+
+        # map averaging type to tuple
+        self._kernel_dict = {'1x1': (1, 1),
```

Comment:
Thinking about this more - why do you not just allow for arbitrary kernel sizes (or at least restrict to odd ones with a 2n+1 mapping). Having the list seems unnecessarily restrictive. I suggest having an `_kernel` property that is either a single integer (in which case it gets copied across all dimensions) or is a tuple (with the same length as dimensions). In the GUI I would then expose a `QSpinBox` like you do for length, and as that is adjusted it adjusts the `_kernel` to be the integer represented their (possibly after the 2n+1 transform), so that the `QSpinBox` would show `0` for no averaging.

Reply:
Exactly. This is one of many spots where I think an nd generalisation would be trivial, and actually result in cleaner code.

Reply:
Ok, I've changed this so that it takes a single QSpinbox and allows arbitrary 2n+1 values.  

It would be nice to have two spinboxes side by side, one for each (x,y) kernel.  But assigning two columns to a single row of grid_layout will alter all other rows.  I tried adding a layout to the specific row of grid_layout, but then it causes "hide" problems ...

Reply:
Nice! My concern about having the two spinboxes is that it won't generalize to ND images. I'd say for now we don't worry about it and try and get this merged, and then we can revisit this when we think about the nD generalization